src/Agent.PluginCore/PluginUtil.cs

```diff
@@ -0,0 +1,1611 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.TeamFoundation.DistributedTask.WebApi;
+using Microsoft.VisualStudio.Services.Common;
+using Microsoft.VisualStudio.Services.OAuth;
+using Microsoft.VisualStudio.Services.WebApi;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Pipelines = Microsoft.TeamFoundation.DistributedTask.Pipelines;
+using Microsoft.Win32;
+
+namespace Agent.PluginCore
+{
+    public class AgentCertificateSettings
+    {
+        public bool SkipServerCertificateValidation { get; set; }
+        public string CACertificateFile { get; set; }
+        public string ClientCertificateFile { get; set; }
+        public string ClientCertificatePrivateKeyFile { get; set; }
+        public string ClientCertificateArchiveFile { get; set; }
+        public string ClientCertificatePassword { get; set; }
+    }
+
+    public class AgentWebProxySettings
+    {
+        public string ProxyAddress { get; set; }
+        public string ProxyUsername { get; set; }
+        public string ProxyPassword { get; set; }
+        public List<string> ProxyBypassList { get; set; }
+
+        private readonly List<Regex> _regExBypassList = new List<Regex>();
+        private bool _initialized = false;
+        public bool IsBypassed(Uri uri)
+        {
+            return string.IsNullOrEmpty(ProxyAddress) || uri.IsLoopback || IsMatchInBypassList(uri);
+        }
+
+        private bool IsMatchInBypassList(Uri input)
```

Comment:
is this bypass stuff intentionally duplicated in commandPlugin.cs?

Reply:
yes, the one in commandplugin.cs is part of the IWebProxy implementation.

Reply:
put the calculation part into pluginutil.cs